### PR TITLE
[RFC] cmake: Pass MIN_LOG_LEVEL as compiler definition

### PR DESCRIPTION
**Pull request by equalsraf**
Allow MIN_LOG_LEVEL to be set as a CMake define.

And maybe put something in local.mk and the wiki, I don't think this is documented anywhere.

### CODE REVIEW

**Comment by Pyrohh** [comment]
Perhaps a `message()` should be printed if it's not explicitly set? Something like "MIN_LOG_LEVEL not explicitly specified, defaulting to INFO_LOG_LEVEL".

Actually, It might be better to extract the `ifndef MIN_LOG_LEVEL` block out of `log.h` and change it to something like this:
```
if(DEFINED MIN_LOG_LEVEL)
  message(STATUS "Log level is set to ${MIN_LOG_LEVEL}")
  add_definitions(-DMIN_LOG_LEVEL=${MIN_LOG_LEVEL})
else()
  add_definitions(-DMIN_LOG_LEVEL=INFO_LOG_LEVEL)
  message(STATUS "Log level not specified, defaulting to ${MIN_LOG_LEVEL}")
endif()
```

File: src/nvim/CMakeLists.txt [hunk 1]
```diff
@@ -80,6 +80,20 @@ endforeach()
 set_source_files_properties(
   ${CONV_SOURCES} PROPERTIES COMPILE_FLAGS "${COMPILE_FLAGS} -Wno-conversion")
 
+if(DEFINED MIN_LOG_LEVEL)
+  if(NOT MIN_LOG_LEVEL MATCHES "^[0-3]$")
+    message(FATAL_ERROR "MIN_LOG_LEVEL must be a number DEBUG(0), INFO(1), WARNING(2), ERROR(3)")
+  endif()
+  message(STATUS "Log level is set to ${MIN_LOG_LEVEL}")
```
Comment (Pyrohh) [inline src/nvim/CMakeLists.txt]: Maybe just `Log level set to ...`.

File: src/nvim/CMakeLists.txt [hunk 2]
```diff
@@ -80,6 +80,20 @@ endforeach()
...
+  add_definitions(-DMIN_LOG_LEVEL=${MIN_LOG_LEVEL})
+else()
+  # By default INFO(1), disabled for Release builds
+  if(NOT CMAKE_BUILD_TYPE STREQUAL "Release")
+    add_definitions(-DMIN_LOG_LEVEL=1)
+    message(STATUS "Log level not specified, defaulting to INFO")
+  endif()
```
Comment (Pyrohh) [inline src/nvim/CMakeLists.txt]: Shouldn't non-release builds be handled here too? Or is that left to the `#ifndef` in log.h on purpose?

Comment (equalsraf) [inline src/nvim/CMakeLists.txt]: I assumed they would be disabled for Release - and log.h has a check for NDEBUG that should cover that case. I didn't want to default to risk a case where logs would be unintentionally disabled.

**Comment by equalsraf** [comment]
IIIRC travis.yml was already setting CFLAGS='-DMIN_LOG_LEVEL=0'

File: src/nvim/CMakeLists.txt (hunk 2 above)
Comment (Pyrohh) [inline src/nvim/CMakeLists.txt]: Sounds fine to me.

**Comment by Pyrohh** [comment]
Would it be possible for this to fail earlier? CMake does a decent amount of configuration before failing on my machine as seen here: https://gist.github.com/Pyrohh/d543656f7e7b09ae2149

----------------------------------------

Also, you might want to add this patch:
```diff
diff --git a/contrib/local.mk.example b/contrib/local.mk.example
index 1c3ea4a..0143e2e 100644
--- a/contrib/local.mk.example
+++ b/contrib/local.mk.example
@@ -24,6 +24,11 @@
 #
 # CMAKE_BUILD_TYPE := Debug
 
+# By default, nvim's log level is INFO (1) (unless CMAKE_BUILD_TYPE is
+# "Release", in which case logging is disabled).
+# The log level must be a number DEBUG (0), INFO (1), WARNING (2), ERROR (3).
+# CMAKE_EXTRA_FLAGS += -DMIN_LOG_LEVEL=1
+
 # By default, nvim uses bundled versions of its required third-party
 # dependencies.
 # Uncomment these entries to instead use system-wide installations of
```
Feel free to change the wording and whatnot. Besides all that, is there anything in particular this PR is missing? I couldn't help but notice the `RFC/WIP` tag.

**Comment by equalsraf** [comment]
> Would it be possible for this to fail earlier? 

I'll try to pull it earlier in CMakeLists.txt

> Besides all that, is there anything in particular this PR is missing? I couldn't help but notice the RFC/WIP tag.

Travis is failing because of the CFLAGS='-DMIN_LOG_LEVEL=0' in travis.yml I'm still thinking on what is the best way to replace it. Worst case scenario I'll accept an environment variable for MIN_LOG_LEVEL, but I would like to avoid it.

**Comment by equalsraf** [comment]
Moved the checks to the main CMakeLists.txt. Default log level is now 0 for travis builds.

File: CMakeLists.txt [hunk 3]
```diff
@@ -285,6 +285,22 @@ endforeach()
 
 install(SCRIPT ${PROJECT_SOURCE_DIR}/cmake/GenerateHelptags.cmake)
 
+# MIN_LOG_LEVEL for log.h
+if(DEFINED MIN_LOG_LEVEL)
+  if(NOT MIN_LOG_LEVEL MATCHES "^[0-3]$")
+    message(FATAL_ERROR "MIN_LOG_LEVEL must be a number DEBUG(0), INFO(1), WARNING(2), ERROR(3)")
```
Comment (Pyrohh) [inline CMakeLists.txt]: To match local.mk (and for easier reading IMO), spaces between the parens (e.g., `WARNING (2)`) might be good,and an `or` before `ERROR`.

File: contrib/local.mk.example [hunk 4]
```diff
@@ -24,6 +24,11 @@
 #
 # CMAKE_BUILD_TYPE := Debug
 
+# By default, nvim's log level is INFO (1) (unless CMAKE_BUILD_TYPE is
+# "Release", in which case logging is disabled).
+# The log level must be a number DEBUG (0), INFO (1), WARNING (2), ERROR (3).
```
Comment (Pyrohh) [inline contrib/local.mk.example]: Same here regarding `or`.

**Comment by Pyrohh** [comment]
Besides the nits (which are pretty subjective), LGTM. What do you think @fwalch?

**Comment by sethjackson** [comment]
@equalsraf can you rebase this? #810 has a similar commit in equalsraf@cc17656 that I was going to open a PR for but this one is already open. :)

**Comment by equalsraf** [comment]
Rebased.

**Comment by sethjackson** [comment]
Thanks! Can someone re-review if necessary and merge? cc @Pyrohh @fwalch.

File: CMakeLists.txt [hunk 5]
```diff
@@ -377,6 +377,23 @@ install_helper(
...
+  endif()
+  message(STATUS "Log level set to ${MIN_LOG_LEVEL}")
+else()
+  # By default INFO(1), or DEBUG(0) in Travis, disabled for Release builds
+  if(TRAVIS_CI_BUILD)
+    message(STATUS "Log level not specified, defaulting to DEBUG(0)")
+    set(MIN_LOG_LEVEL 0)
+  elseif(NOT CMAKE_BUILD_TYPE STREQUAL "Release")
+    message(STATUS "Log level not specified, defaulting to INFO(1)")
+    set(MIN_LOG_LEVEL 1)
```
Comment (justinmk) [inline CMakeLists.txt]: This elseif() isn't needed because log.h already defaults to INFO.

File: CMakeLists.txt [hunk 6]
```diff
@@ -381,6 +381,20 @@ install_helper(
   FILES ${MANPAGES}
   DESTINATION ${CMAKE_INSTALL_MANDIR}/man1)
 
+# MIN_LOG_LEVEL for log.h
+if(DEFINED MIN_LOG_LEVEL)
+  if(NOT MIN_LOG_LEVEL MATCHES "^[0-3]$")
+    message(FATAL_ERROR "MIN_LOG_LEVEL must be a number DEBUG (0), INFO (1), WARNING (2) or ERROR (3)")
+  endif()
+  message(STATUS "Log level set to ${MIN_LOG_LEVEL}")
+elseif(TRAVIS_CI_BUILD)
```
Comment (justinmk) [inline CMakeLists.txt]: Why move this from `.travis.yml`? I think we should avoid sprinkling CI-specific conditions in the build files--try to keep it in .travis.yml.

File: .travis.yml [hunk 7]
```diff
@@ -12,8 +12,6 @@ env:
     - PATH="$(python2.7 -c 'import site; print(site.getuserbase())')/bin:$HOME/.local/mingw32/bin:$PATH"
     # LLVM symbolizer path.
     - LLVM_SYMBOLIZER="$(which llvm-symbolizer-3.6)"
-    # Force verification of DLOG macros.
-    - CFLAGS="-DMIN_LOG_LEVEL=0"
```
Comment (justinmk) [inline .travis.yml]: can this be done in CMAKE_FLAGS on line 26 now?

File: CMakeLists.txt (hunk 6 above)
Comment (equalsraf) [inline CMakeLists.txt]: Removed TRAVIS specific elseif.

File: .travis.yml (hunk 7 above)
Comment (equalsraf) [inline .travis.yml]: Yes, moved it to line 26

**Comment by justinmk** [comment]
@equalsraf The failing test isn't related to this PR, I will merge this now.

**Comment by equalsraf** [comment]
Ups just pushed a squashed commit.

**Pull request merged**